### PR TITLE
[#436] feat(client,server): Introduce multi-part LocalStorageManager

### PR DESCRIPTION
**Pull request by maobaolong**
### What changes were proposed in this pull request?

- Introduce a factory to create specific LocalStorageManager by config.
- Introduce multiply disk LocalStorageManager.

### Why are the changes needed?

Fix: #436 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

- Existing UTs and new added UT
- Tested on our pressure test cluster.
  - new client -> old server ✓
  - new client -> new server ✓
  - old client -> new server ❌, so we have to upgraded client first, than upgrade the servers

### CODE REVIEW

**Comment by maobaolong** [comment]
@zuston Thank you advance for your previous work on #436 , would you like to take a look at this PR?

**Comment by zuston** [comment]
> @zuston Thank you advance for your previous work on #436 , would you like to take a look at this PR?

Yes. I will take a look in the next 3 days.

**Comment by zuston** [comment]
Thanks for proposing this. But I'm afraid of the implicit storage index behind the offset, which is dangerous. Why not introducing the extra storage index into the protobuf.

**Comment by maobaolong** [comment]
@zuston Thanks for your review, in fact I have done another approach with the proto change, which introduce storageId into the shuffleIndex file and changed the netty encode/decode methods in the client/server side, it can be more graceful than this PR and it can works as expected too. But I cannot give a gray online plan, I cannot upgrade client and server at this same time.

For this PR, although both client and server are need to upgrade, but I can upgrade all the clients first, then I upgrade the server.

**Comment by zuston** [comment]
> Thanks for your review, in fact I have done another approach with the proto change, which introduce storageId into the shuffleIndex file and changed the netty encode/decode methods in the client/server side

I don't think the storage id should be recorded into the index file, this could be retrieved from the local storage index. If so, the client could do some compatible change for this.

**Comment by maobaolong** [comment]
@zuston Thanks for the discussion and suggestion, I change to add storageIds as the response of `getShuffleIndex`, it looks clear than the previous tricky approach which use `high 8 bits` of offset as storageId, and the newer approach did not modify the layout of index file in server, this is a good idea, PTAL.

**Review by zuston — COMMENTED** [review]
Overall LGTM now. Thanks for your effort @maobaolong

**Comment by zuston** [comment]
BTW, `SHUFFLE_SERVER_VERSION` should be updated, because the new server is not compatible with the old version client.  @maobaolong

File: server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java [hunk 1]
```diff
@@ -744,6 +745,12 @@ public class ShuffleServerConf extends RssBaseConf {
           .defaultValue(false)
           .withDescription("Whether to enable app detail log");
 
+  public static final ConfigOption<String> SERVER_LOCAL_STORAGE_MANAGER_CLASS =
+      ConfigOptions.key("rss.server.localStorageManagerClass")
+          .stringType()
+          .defaultValue(MultiPartLocalStorageManager.class.getName())
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply server/src/main/java/org/apache/uniffle/server/ShuffleServerConf.java:751]: Sure, will revert it

File: server/src/main/java/org/apache/uniffle/server/storage/MultiPartLocalStorageManager.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    // Use higher 8 bit to storage the storage id, and use lower 56 bit to storage the offset.
+    int storageId = event.getStorageId();
+    // TODO(baoloongmao): check AOOB exception
+    return idToStorages.get(storageId);
+  }
+
+  @Override
+  public Storage selectStorageForIndex(ShuffleDataReadEvent event) {
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply server/src/main/java/org/apache/uniffle/server/storage/MultiPartLocalStorageManager.java:116]: Done

File: storage/src/main/java/org/apache/uniffle/storage/common/CompositeStorage.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import org.slf4j.LoggerFactory;
+
+import org.apache.uniffle.common.exception.FileNotFoundException;
+import org.apache.uniffle.storage.handler.api.ServerReadHandler;
+import org.apache.uniffle.storage.handler.api.ShuffleWriteHandler;
+import org.apache.uniffle.storage.handler.impl.CompositeLocalFileServerReadHandler;
+import org.apache.uniffle.storage.request.CreateShuffleReadHandlerRequest;
+import org.apache.uniffle.storage.request.CreateShuffleWriteHandlerRequest;
+
+public class CompositeStorage extends AbstractStorage {
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply storage/src/main/java/org/apache/uniffle/storage/common/CompositeStorage.java:33]: Done

File: common/src/main/java/org/apache/uniffle/common/segment/FixedSizeSegmentSplitter.java [hunk 4]
```diff
@@ -91,24 +102,35 @@ private static List<ShuffleDataSegment> transIndexDataToSegments(
           break;
         }
 
-        bufferSegments.add(
-            new BufferSegment(blockId, bufferOffset, length, uncompressLength, crc, taskAttemptId));
-        bufferOffset += length;
+        boolean storageChanged = preStorageId != -1 && currentStorageId != preStorageId;
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply common/src/main/java/org/apache/uniffle/common/segment/FixedSizeSegmentSplitter.java:105]: Done

File: common/src/main/java/org/apache/uniffle/common/segment/LocalOrderSegmentSplitter.java [hunk 5]
```diff
@@ -112,27 +130,34 @@ public List<ShuffleDataSegment> split(ShuffleIndexResult shuffleIndexResult) {
...
             lastExpectedBlockIndex != -1
                 && bufferSegments.size() > 0
                 && expectTaskIds.contains(taskAttemptId)
                 && index - lastExpectedBlockIndex != 1;
 
         boolean conditionOfLimitedBufferSize = bufferOffset >= readBufferSize;
-
-        if (conditionOfDiscontinuousBlocks || conditionOfLimitedBufferSize) {
-          ShuffleDataSegment sds = new ShuffleDataSegment(fileOffset, bufferOffset, bufferSegments);
+        boolean storageChanged =
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply common/src/main/java/org/apache/uniffle/common/segment/LocalOrderSegmentSplitter.java:140]: Done

File: common/src/main/java/org/apache/uniffle/common/segment/LocalOrderSegmentSplitter.java [hunk 6]
```diff
@@ -85,7 +85,12 @@ public List<ShuffleDataSegment> split(ShuffleIndexResult shuffleIndexResult) {
      * ShuffleDataSegment size should < readBufferSize 3. Single shuffleDataSegment's blocks should
      * be continuous
      */
+    int[] storageIds = shuffleIndexResult.getStorageIds();
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply common/src/main/java/org/apache/uniffle/common/segment/LocalOrderSegmentSplitter.java]: Done

**Comment by maobaolong** [comment]
@zuston All the suggestion has been addressed. PTAL

File: common/src/main/java/org/apache/uniffle/common/segment/LocalOrderSegmentSplitter.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (zuston) [reply common/src/main/java/org/apache/uniffle/common/segment/LocalOrderSegmentSplitter.java]: I haven't seen the newly test cases about this class.

[the remark this answers is not in the document]
Comment (zuston) [reply common/src/main/java/org/apache/uniffle/common/segment/LocalOrderSegmentSplitter.java]: My fault, catch it.

**Review by zuston — DISMISSED** [review]
+1. I hope this feature could be described in the doc in the another PR.

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/CompositeFileRegion.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+ */
+
+package org.apache.uniffle.common.netty.protocol;
+
+import java.io.IOException;
+import java.nio.channels.WritableByteChannel;
+
+import io.netty.channel.FileRegion;
+
+public class CompositeFileRegion extends AbstractFileRegion {
```
Comment (jerqi) [inline common/src/main/java/org/apache/uniffle/common/netty/protocol/CompositeFileRegion.java:25]: @rickyma Could you help review the `Netty` part?

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/CompositeFileRegion.java [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    super.retain(increment);
+    for (FileRegion region : regions) {
+      region.retain(increment);
+    }
+    return this;
+  }
+
+  @Override
+  public boolean release() {
+    super.release();
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/CompositeFileRegion.java:94]: Done

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/CompositeFileRegion.java [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    for (FileRegion region : regions) {
+      if (!region.release(decrement)) {
+        released = false;
+      }
+    }
+    return released;
+  }
+
+  @Override
+  protected void deallocate() {}
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/CompositeFileRegion.java:117]: Added

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/CompositeFileRegion.java [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  @Override
+  public long transferTo(WritableByteChannel target, long position) throws IOException {
+    long totalTransferred = 0;
+
+    for (FileRegion region : regions) {
+      if (position >= region.count()) {
+        position -= region.count();
+      } else {
+        long transferredNow = region.transferTo(target, position);
+        totalTransferred += transferredNow;
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/CompositeFileRegion.java:56]: done

**Comment by maobaolong** [comment]
@rickyma Thanks for your review and suggestion. It helps to make this PR better.

I updated the code and addressed your comments, PTAL.

File: common/src/main/java/org/apache/uniffle/common/netty/protocol/CompositeFileRegion.java [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+      if (!region.release()) {
+        released = false;
+      }
+    }
+    return released;
+  }
+
+  @Override
+  public boolean release(int decrement) {
+    super.release(decrement);
```
[the remark this answers is not in the document]
Comment (maobaolong) [reply common/src/main/java/org/apache/uniffle/common/netty/protocol/CompositeFileRegion.java]: done

**Review by jerqi — DISMISSED** [review]

**Comment by maobaolong** [comment]
@jerqi @rickyma @zuston Thanks for your review, will merge after our final test pass.

**Review by jerqi — APPROVED** [review]

**Comment by maobaolong** [comment]
@rickyma The title has been updated.

**Review by zuston — APPROVED** [review]

**Pull request merged**